File: packages/engine/src/validation.ts
import {
  type BattleInput,
  type EffectTemplateInput,
  ROW_TYPES,
  type RowType,
  STAT_KEYS,
  TARGET_PRIORITIES,
  TARGET_SCOPES,
  TARGET_SELECTION_SHAPES,
  type UnitInput,
} from "./types";

export class InvalidBattleInputError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InvalidBattleInputError";
  }
}

export type InvalidBattleStateCode = "CASTER_SCENARIO_NOT_FOUND" | "OPPOSING_SCENARIO_NOT_FOUND";

export class InvalidBattleStateError extends Error {
  readonly code: InvalidBattleStateCode;

  constructor(code: InvalidBattleStateCode, message: string) {
    super(message);
    this.name = "InvalidBattleStateError";
    this.code = code;
  }
}

export function validateBattleInput(input: BattleInput): void {
  if (typeof input.seed !== "number" && typeof input.seed !== "string") {
    throw new InvalidBattleInputError("Battle seed must be a finite number or non-blank string.");
  }
  if (typeof input.seed === "number" && !Number.isFinite(input.seed)) {
    throw new InvalidBattleInputError("Battle seed must be a finite number or non-blank string.");
  }
  if (typeof input.seed === "string" && input.seed.trim().length === 0) {
    throw new InvalidBattleInputError("Battle seed must not be blank.");
  }

  const deployedUnits = input.scenarios.flatMap((scenario) =>
    ROW_TYPES.flatMap((rowType) =>
      (scenario.rows?.[rowType] ?? []).map((unit) => ({ rowType, unit })),
    ),
  );
  for (const { rowType, unit } of deployedUnits) {
    validateTargetingConfiguration(unit);
    validateTargetCount(unit);
    validateItemRows(unit, rowType);
    validateEffectTiming(unit);
  }

  const livingUnits = deployedUnits
    .map(({ unit }) => unit)
    .filter((unit) => (unit.currentHealth ?? unit.stats.health) > 0);

  if (livingUnits.length === 0) {
    throw new InvalidBattleInputError("Battle initialization requires at least one living unit.");
  }
}

function validateEffectTiming(unit: UnitInput): void {
  for (const { effect } of (unit.items ?? []).flatMap((item) => item.effects ?? [])) {
    validateEffectShieldLifecycle(effect);

    if (effect.isTaunt && effect.timingType === "interval") {
      throw new InvalidBattleInputError(
        `Taunt effect "${effect.name ?? "Effect"}" must use instant timing.`,
      );
    }
    if (
      effect.isTaunt &&
      effect.timingType === "instant" &&
      effect.lastsForActions != null &&
      (!Number.isInteger(effect.lastsForActions) || effect.lastsForActions <= 0)
    ) {
      throw new InvalidBattleInputError(
        `Taunt effect "${effect.name ?? "Effect"}" must have a positive duration.`,
      );
    }

    const intervalTimingMissing =
      effect.timingType === "interval" &&
      (effect.triggerEveryActions == null || effect.triggerCount == null);
    const hasStatModifier = STAT_KEYS.some((statKey) => effect[statKey] != null);
    const hasShield = typeof effect.shield === "number" && effect.shield > 0;
    const actionDurationMissing =
      !effect.isTaunt &&
      effect.timingType === "instant" &&
      (effect.effectType === "buff" || effect.effectType === "debuff") &&
      (hasStatModifier || hasShield) &&
      effect.lastsForActions == null;

    if (intervalTimingMissing || actionDurationMissing) {
      throw new InvalidBattleInputError(
        `Effect "${effect.name ?? "Effect"}" timing needs configuration.`,
      );
    }
  }
}

export function validateEffectShieldLifecycle(effect: EffectTemplateInput): void {
  const hasPositiveShield = typeof effect.shield === "number" && effect.shield > 0;
  const hasSupportedLifecycle =
    effect.timingType === "instant" &&
    (effect.effectType === "buff" || effect.effectType === "debuff");

  if (hasPositiveShield && !hasSupportedLifecycle) {
    throw new InvalidBattleInputError(
      `Effect "${effect.name ?? "Effect"}" shield is only supported for instant buffs and debuffs.`,
    );
  }
}

function validateTargetingConfiguration(unit: UnitInput): void {
  if (
    unit.targetScope !== undefined &&
    !TARGET_SCOPES.some((targetScope) => targetScope === unit.targetScope)
  ) {
    throw new InvalidBattleInputError(
      `Invalid target scope "${String(unit.targetScope)}" for ${unit.name}.`,
    );
  }
  if (
    unit.targetPriority !== undefined &&
    !TARGET_PRIORITIES.some((targetPriority) => targetPriority === unit.targetPriority)
  ) {
    throw new InvalidBattleInputError(
      `Invalid target priority "${String(unit.targetPriority)}" for ${unit.name}.`,
    );
  }
  if (
    unit.selectionShape !== undefined &&
    !TARGET_SELECTION_SHAPES.some((selectionShape) => selectionShape === unit.selectionShape)
  ) {
    throw new InvalidBattleInputError(
      `Invalid selection shape "${String(unit.selectionShape)}" for ${unit.name}.`,
    );
  }
}

function validateTargetCount(unit: UnitInput): void {
  if (
    unit.targetCount !== undefined &&
    (!Number.isInteger(unit.targetCount) || unit.targetCount <= 0)
  ) {
    throw new InvalidBattleInputError("Target count must be a positive integer");
  }
}

function validateItemRows(unit: UnitInput, deployedRow: RowType): void {
  let allowedRows = new Set<RowType>(ROW_TYPES);

  for (const item of unit.items ?? []) {
    const itemRows = item.allowedRowTypes ?? [];
    if (new Set(itemRows).size !== itemRows.length) {
      throw new InvalidBattleInputError(`${item.name} has duplicate allowed row types`);
    }

    const restriction = itemRows.length === 0 ? ROW_TYPES : itemRows;
    allowedRows = new Set([...allowedRows].filter((rowType) => restriction.includes(rowType)));
  }

  if (allowedRows.size === 0) {
    throw new InvalidBattleInputError(`${unit.name} has no shared allowed item rows`);
  }

  if (!allowedRows.has(deployedRow)) {
    throw new InvalidBattleInputError(`${unit.name} cannot be deployed in ${deployedRow}`);
  }
}

File: crates/qd-engine/src/validation.rs
use crate::model::*;
use serde_json::Value;
fn fail(message: impl Into<String>) -> Result<(), EngineError> {
    Err(EngineError(message.into()))
}
pub(crate) fn validate(input: &Value) -> Result<(), EngineError> {
    canonical_seed(&input["seed"])?;
    if input["scenarios"].as_array().map(Vec::len) != Some(2) {
        return fail("Battle initialization requires exactly two scenarios.");
    }
    let scenarios = array(input, "scenarios");
    let mut alive = false;
    for scenario in scenarios {
        if scenario["id"].as_str().is_none() {
            return fail("Scenario id must be a string.");
        }
        for row in ROWS {
            for unit in array(&scenario["rows"], row) {
                for (key, label, allowed) in [
                    (
                        "targetScope",
                        "target scope",
                        &[
                            "self",
                            "self_allies",
                            "self_enemies",
                            "allies",
                            "enemies",
                            "both",
                        ][..],
                    ),
                    (
                        "targetPriority",
                        "target priority",
                        &[
                            "highest_health",
                            "lowest_health",
                            "highest_damage",
                            "support",
                            "random",
                        ][..],
                    ),
                    (
                        "selectionShape",
                        "selection shape",
                        &["individual", "adjacent"][..],
                    ),
                ] {
                    if let Some(v) = unit.get(key)
                        && !v.as_str().is_some_and(|v| allowed.contains(&v))
                    {
                        let text = v
                            .as_str()
                            .map(str::to_owned)
                            .unwrap_or_else(|| v.to_string());
                        return fail(format!(
                            "Invalid {label} \"{text}\" for {}.",
                            string(&unit, "name")
                        ));
                    }
                }
                if let Some(count) = unit.get("targetCount")
                    && !count.as_f64().is_some_and(|n| n > 0.0 && n.fract() == 0.0)
                {
                    return fail("Target count must be a positive integer");
                }
                let items = array(&unit, "items");
                let mut allowed = ROWS.to_vec();
                for item in &items {
                    let row_values = array(item, "allowedRowTypes");
                    let item_rows: Vec<&str> = row_values
                        .iter()
                        .map(|v| v.as_str().unwrap_or(""))
                        .collect();
                    let unique: std::collections::HashSet<_> = item_rows.iter().collect();
                    if unique.len() != item_rows.len() {
                        return fail(format!(
                            "{} has duplicate allowed row types",
                            string(item, "name")
                        ));
                    }
                    if !item_rows.is_empty() {
                        allowed.retain(|r| item_rows.contains(r));
                    }
                }
                if allowed.is_empty() {
                    return fail(format!(
                        "{} has no shared allowed item rows",
                        string(&unit, "name")
                    ));
                }
                if !allowed.contains(&row) {
                    return fail(format!(
                        "{} cannot be deployed in {row}",
                        string(&unit, "name")
                    ));
                }
                for item in &items {
                    for sequence in array(item, "effects") {
                        let e = &sequence["effect"];
                        let timing = string(e, "timingType");
                        let category = string(e, "effectType");
                        let taunt = e["isTaunt"] == true;
                        let modifiers = STATS.iter().any(|key| !e[*key].is_null());
                        let shield = number(e, "shield") > 0.0;
                        if shield
                            && !(timing == "instant"
                                && (category == "buff" || category == "debuff"))
                        {
                            return fail(format!(
                                "Effect \"{}\" shield is only supported for instant buffs and debuffs.",
                                name(e)
                            ));
                        }
                        if taunt && timing == "interval" {
                            return fail(format!(
                                "Taunt effect \"{}\" must use instant timing.",
                                name(e)
                            ));
                        }
                        if taunt
                            && timing == "instant"
                            && !e["lastsForActions"].is_null()
                            && !e["lastsForActions"]
                                .as_f64()
                                .is_some_and(|n| n > 0.0 && n.fract() == 0.0)
                        {
                            return fail(format!(
                                "Taunt effect \"{}\" must have a positive duration.",
                                name(e)
                            ));
                        }
                        if (timing == "interval"
                            && (e["triggerEveryActions"].is_null() || e["triggerCount"].is_null()))
                            || (!taunt
                                && timing == "instant"
                                && (category == "buff" || category == "debuff")
                                && (modifiers || shield)
                                && e["lastsForActions"].is_null())
                        {
                            return fail(format!(
                                "Effect \"{}\" timing needs configuration.",
                                name(e)
                            ));
                        }
                    }
                }
                if optional_number(&unit, "currentHealth", number(&unit["stats"], "health")) > 0.0 {
                    alive = true;
                }
            }
        }
    }
    if !alive {
        return fail("Battle initialization requires at least one living unit.");
    }
    Ok(())
}
